General video component

deaxmachina opened the pull request:
### General-purpose video component. 
- Can play on load or when the video comes into view with intersection observer.
- Supports videos with or without audio.
- Has optional controls for play/pause/replay with some optional styling, and option whether to allow the video to be played/paused at all. 
- Has basic accessibility option for screen readers. 

We've used videos more often in projects recently and this component aims to make it easier to add video to pages as a component similar to the Image one. It should cover the most common use-cases.

hobbes7878 commented:
This is a really cool component, @deaxmachina. 

I just took a quick glance at it, loved it and had just a couple notes:

- I would expect the `src` component to come in already prepended with `assets` rather than doing that within the component. That makes sure this component works if the video is hosted outside the project. E.g., one day we may have a video optimiser microservice that puts optimised videos somewhere else on a cdn or whatever...
- I might allow audio videos the option to autoplay, rather than hard code against it. Use case I'm thinking of is if you have multiple videos on the page or some other way you want to control for that first interaction, say outside this component. Right now, it's a hard stop even if you somehow know that somewhere else on the page you've gotten the interaction you need to allow autoplay audio vids in the browser. The note you have explaining the reasons _generally_ not to do it is perfect; I just wouldn't eliminate the possibility if someone really does know what they're doing on a more complex page.
- Might be nice to have an option where the entire video can be a touchpoint for interacting with the controls, rather than just the button itself. That's handy for mobile, especially.
- Small typo in the props, I think: separateRepl**a**yIcon

That's all my thoughts at first glance.

He's the busiest guy on Reuters Graphics right now, but if @fcage gets a chance to look, I think he might be better to comment on how well this fits our most common video treatments in pages.

Other than those, I'm happy to merge this and publish the component.

deaxmachina commented:
Thank you, @hobbes7878 ! 

(1) By "`src` component to come in already prepended with `assets`", do you mean `assets` should be passed explicitly as part of the path in the prop by the user? I can do that. 
(2) Yeah, that makes sense. I was just getting errors in the console when `muted` was set to `false` and there was no interaction. I'm not sure if console errors are a deal breaker, or else if I can circumvent it another way. I will give it a try tomorrow. 
(3) If I understand correctly, I think it can already do that. With the `showControls={false}` option (while `possibleToPlayPause={true}` which is by default)

Btw I'm not sure if something has changed in the way replying to comments works here, but the quote-reply looked strange, so replying with an @

hobbes7878 commented:
You can always @ me! :) 

1. Exactly. So a user would do something like:

```svelte
<Video
  src={`${assets}/videos/myvid.mp4`}
/>
```
2. Console logs are OK as long as they're warning useful things! In this case, especially, I think it's good to show an error if not handled correctly.
3. OK, cool cool. I missed that.

Eventually, I wanna get other folks to look at this, but maybe we'll go ahead and publish with these changes tomorrow, you can write it up in global graphics and then we can adjust to anything good folks raise in the next point release.

fcage commented:
Yeah looks good to me. Few thoughts below though no "issues".

Two things I saw have already been covered: passing full source to component so it can come from anywhere, and autoplay as a prop. If we are passing the full source url, can we remove the `  import { assets } from '$app/paths';` reference since it's not needed? I ask because this becomes universal for the old rig too which unfortunately we will use for sreps for the time being.

Do we want to pass the option of a poster? I'm conflicted but not sure how all browsers respond if we just preload metadata. 

Thinking aloud as a nice to have. Is the option for the controls show/dont show? Or would we want them to show when paused, or as you scroll into view and it plays, and they hide until hover?

deaxmachina commented:
@hobbes7878 , @fcage  Thanks, both. Added the following:
1. Reference to `assets` removed from component, added docs for that. 
2. Added option to show/hide the controls on hover with `hoverToSeeControls` 
3. For videos with audio, added prop `allowSoundToAutoplay` which, if set to `true` allows for a video with sound to autoplay if the user has interacted with the page. No console errors either; fyi, all it took was 
```  
let interactedWithDom = false;
const setInteractedWithDom = () => { interactedWithDom = true; }; 

<svelte:window on:click={setInteractedWithDom} on:touchstart={setInteractedWithDom}/>
```
to check for DOM interactions, and then a bunch more conditionals. Since there are now quite a few permutations on how the component might behave, I'll keep an eye out for bugs in the wild. I tested, but might have missed some combination. 
4. Fixed a bug on the 'invisible' controls, it should now work ok to click on the video to play/pause with the controls hidden option.

hobbes7878 commented:
released as `0.3.12`
